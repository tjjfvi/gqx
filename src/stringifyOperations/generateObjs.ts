
import { Context, Obj, Id, Prop } from ".";
import unwrapType from "../unwrapType";

export default (ctx: Context): [Obj[], Context] => {
  const objs: Obj[] =
    ctx.operations
      .map(([prop, name]) => {
        const { fields } = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          const [{ name: { value: typeName } }, wrap] = unwrapType(field.type);
          let prop: Prop = {
            id,
            type: typeName,
            wrap,
            args: field.arguments?.slice() ?? [],
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const obj: Obj = {
          type: name,
          prop,
          props,
        };
        return obj;
      })
  return [objs, ctx];
}
