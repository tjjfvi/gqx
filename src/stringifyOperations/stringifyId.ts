
import { Id } from ".";

export const stringifyId = (id: Id) =>
  `"${id.type}$${id.prop}"`
