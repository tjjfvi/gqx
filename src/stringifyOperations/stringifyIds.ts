
import { Id } from ".";
import stringifyId from "./stringifyId";

export default (ids: Id[]) =>
  ids.map(id => `class ${stringifyId(id)} { declare private __nominal__: any; }\n`).join("")
