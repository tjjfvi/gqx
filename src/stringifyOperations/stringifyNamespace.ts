import { Obj } from "."
import stringifyId from "./stringifyId"
import { Prop } from ".";

export default (
  name: string,
  objs: Obj[],
  f: (obj: Obj, prop: Prop) => string,
  extraGenerics = "",
  cond = (o: Obj, p: Prop) =>
    `    T extends typeof ${stringifyId(p.id)} ? ${o.type}.${p.id.prop} :`
) =>
  `
export namespace ${name} {
${
  objs.map(o =>
    `  export namespace ${o.type} {\n${o.props.map(p =>
      `    export ` + f(o, p) + `\n`
    ).join("")}  }\n`
  ).join("")
}
  export type $<T${extraGenerics}> =
${
  objs.map(o => o.props.map(p =>
    cond(o, p)
  ).join("\n")).join("\n")
}
    never
}
`.trim()
