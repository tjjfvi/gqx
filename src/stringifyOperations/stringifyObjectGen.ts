import { indent } from "../indent"
import { Obj } from "."
import { stringifyId } from "./stringifyId"

export const stringifyObjectGen = (objs: Obj[]) =>
  `
export type $OperationTypes = ${objs.map(o => JSON.stringify(o.type)).join(" | ")}
export const $gqx = <F extends $Hkt<$OperationProp>>(f: <P extends $OperationProp>(p: P) => $Hkt.Call<F, P>) => ({\n${
    objs.map(o =>
      indent`${o.prop}: {\n${
        o.props.map(p => indent`${p.id.prop}: f(${stringifyId(p.id)}),\n`).join("")
      }},\n`
    ).join("")
  }})
`.trim()
