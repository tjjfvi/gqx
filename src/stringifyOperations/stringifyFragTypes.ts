
import { Obj, Context } from ".";
import stringifyNamespace from "./stringifyNamespace";

export default (objs: Obj[], ctx: Context) =>
  stringifyNamespace("$$Frag", objs, (o, p) =>
    `type ${p.id.prop} = $$DeepArray<${(p.type in ctx.objectTypes ? `${p.type}$` : "never")}>;`
  )
