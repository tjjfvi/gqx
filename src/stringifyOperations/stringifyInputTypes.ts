
import { Obj } from ".";
import stringifyInputType from "../stringifyInputType";
import stringifyNamespace from "./stringifyNamespace";

export default (objs: Obj[]) =>
  stringifyNamespace("$$Input", objs, (o, p) => stringifyInputType(p.id.prop, p.args))
