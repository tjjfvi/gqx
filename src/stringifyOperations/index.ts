
import { Context } from "..";
import unwrapType from "../unwrapType";
import { InputValueDefinitionNode } from "graphql";
import stringifyProps from "./stringifyProps"
import stringifyInputTypes from "./stringifyInputTypes";
import stringifyFragTypes from "./stringifyFragTypes";
import stringifyOutputTypes from "./stringifyOutputTypes";
import stringifyVariables from "./stringifyVariables";


interface Id {
  type: string;
  prop: string;
}

interface Prop {
  id: Id;
  type: string;
  wrap: (x: string) => string;
  args: InputValueDefinitionNode[];
}

interface Obj {
  type: string;
  prop: string;
  props: Prop[];
}

export { Id, Prop, Obj, Context };

export default (ctx: Context) => {
  const ids: Id[] = [];

  const objs: Obj[] =
    ctx.operations
      .map(([prop, name]) => {
        const fields = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          ids.push(id);
          const [{ name: { value: typeName } }, wrap] = unwrapType(field.type);
          let prop: Prop = {
            id,
            type: typeName,
            wrap,
            args: field.arguments.slice(),
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const obj: Obj = {
          type: name,
          prop,
          props,
        };
        return obj;
      })

  return [
    stringifyProps(objs),
    stringifyInputTypes(objs),
    stringifyFragTypes(objs, ctx),
    stringifyOutputTypes(objs, ctx),
    stringifyVariables(objs, ctx),
  ].join("\n\n");
}

