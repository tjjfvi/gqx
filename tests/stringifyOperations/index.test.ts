import stringifyOperations from "../../src/stringifyOperations";

test("", () => {
  const ctx = utils.getOperations(["extensive"]);
  const code = stringifyOperations(ctx);
  expect(code).toParse().toMatchSnapshot();
})
