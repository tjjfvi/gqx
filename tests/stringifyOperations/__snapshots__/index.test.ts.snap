// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
"class Query$getAuthor {
  private static _: any;
  static typeProp = \\"query\\";
  static type = \\"Query\\";
  static prop = \\"getAuthor\\";
  static inputTypes = {
    id: \\"ID!\\",
  };
}

class Query$getBook {
  private static _: any;
  static typeProp = \\"query\\";
  static type = \\"Query\\";
  static prop = \\"getBook\\";
  static inputTypes = {
    id: \\"ID!\\",
  };
}

class Query$listBooks {
  private static _: any;
  static typeProp = \\"query\\";
  static type = \\"Query\\";
  static prop = \\"listBooks\\";
  static inputTypes = {
    cursor: \\"Cursor\\",
    filter: \\"BookFilter\\",
  };
}

class Mutation$addReview {
  private static _: any;
  static typeProp = \\"mutation\\";
  static type = \\"Mutation\\";
  static prop = \\"addReview\\";
  static inputTypes = {
    input: \\"ReviewInput!\\",
  };
}



export namespace $$Input {
  export namespace Query {
    export interface getAuthor {
      id: ID;
    }
    export interface getBook {
      id: ID;
    }
    export interface listBooks {
      cursor?: (Cursor | null | void);
      filter?: (BookFilter | null | void);
    }
  }
  export namespace Mutation {
    export interface addReview {
      input: ReviewInput;
    }
  }
  export type $<T> =
    T extends typeof Query$getAuthor ? Query.getAuthor :
    T extends typeof Query$getBook ? Query.getBook :
    T extends typeof Query$listBooks ? Query.listBooks :
    T extends typeof Mutation$addReview ? Mutation.addReview :
    never
}

export namespace $$Frag {
  export namespace Query {
    export type getAuthor = Author$;
    export type getBook = Book$;
    export type listBooks = Book$;
  }
  export namespace Mutation {
    export type addReview = Review$;
  }
  export type $<T> =
    T extends typeof Query$getAuthor ? Query.getAuthor :
    T extends typeof Query$getBook ? Query.getBook :
    T extends typeof Query$listBooks ? Query.listBooks :
    T extends typeof Mutation$addReview ? Mutation.addReview :
    never
}

export namespace $$Output {
  export namespace Query {
    export type getAuthor<F extends $$Frag.$<typeof Query$getAuthor>[]> = $Author<F>
    export type getBook<F extends $$Frag.$<typeof Query$getBook>[]> = $Book<F>
    export type listBooks<F extends $$Frag.$<typeof Query$listBooks>[]> = $Book<F>[]
  }
  export namespace Mutation {
    export type addReview<F extends $$Frag.$<typeof Mutation$addReview>[]> = $Review<F>
  }
  export type $<T, F extends $$Frag.$<T>[]> =
    F extends $$Frag.$<typeof Query$getAuthor>[] ? Query.getAuthor<F> :
    F extends $$Frag.$<typeof Query$getBook>[] ? Query.getBook<F> :
    F extends $$Frag.$<typeof Query$listBooks>[] ? Query.listBooks<F> :
    F extends $$Frag.$<typeof Mutation$addReview>[] ? Mutation.addReview<F> :
    never
}

"
`;
